### ISSUE TITLE
Add the ability to resume an upload.

### ISSUE THREAD

**Issue by fr4gment** [Velocidex/velociraptor#3698]
This is a feature request. 

There are instances where (due to many uncontrollable factors) that an upload is interrupted. Normally for small file uploads, this is fine as we can just re-run the artifact and the upload will finish quickly. If we need to upload a large file (> 1 GB), the client has some mechanisms to handle interrupts (a checkpoint file is created) but if for some reason the collection was stopped (e.g. usually client crashing, flow times out, etc ) then the collection needs to run again and it wastes precious time as the entire upload starts over.

Having the ability to resume or start a collection from where it last left off would be very beneficial in these situations. I realize this may be difficult to do for all functions/plugins, but I think allowing it (somehow) for the upload function would add a lot of value.

**Comment by scudette** [comment]
This is an interesting idea but it has some edge cases to work out. The reason we have timeouts and upload limits is because sometimes an artifact accidentally tries to upload too much data - for example running something like KapeFiles.Targets uses various globs to select the files to upload and sometimes there is just too much data there. I have seen cases where a simple target like "User's temporary internet files" ended up being hundreds of Gb because this happened to be a terminal server with hundreds of users!

When a client crashes or an artifact hits its timeout or upload limit this means that something unexpected happened - that is why we stop and force the user to rethink their collections strategy. If we just automatically tried to resume the upload we will be making matters worse by doubling down on a mistake.

One way to do what you suggest is to create the concept of a "upload transaction" where the artifact can register it wants a file uploaded but then the upload happens in another process independent of the actual artifact completing. Then if the query itself times out the upload continue (with resuming) in the background.

This design allows for resuming even if the client crashes or an artifact is cancelled but then we have the opposite problem: How do we stop the files from uploading? I mean it may be possible to start hundreds of "upload transactions" so we are lining up many Gb in future uploads. This will block up the client's bandwidth for days! and now we have to manage the upload lifetime in a separate concept - i.e. cancelling the query will not automatically stop the uploads!

This has a large potential for major accidents that can bring the server down as well.

So currently the upload lifetime is tied to the query lifetime - if the query dies, the upload stops. This is fairly safe as a simple reboot can reset the system.

You suggestion makes sense for artifacts that upload a lot of files like KapeFiles.Targets because the same query can be responsible for hundreds of uploads - if the query is interrupted then the whole thing needs to be retried. 

Previously we had issues because hunts could schedule many artifacts and they were all issued to the client at once when it checked in. This means there could be say 10-20 queries lined up on the same client. While client side they do not all run at the same time, these queries are waiting to run on the client. This means that if the client crashes say after 3 artifacts, the other 17 collections fail. 

In 0.73 this situation is improved where the server will not issue more than 2 collections at the same time to the client, so if the client crashes we can resume the collections which it has not done yet. This means if 20 collections are lined up and the client crashes after 3 then we lose the first 4 collections but continue from collection 5-20.

The reason I mention this in this thread is because this strategy makes it more reliable to split upload into multiple collections, where each collection is responsible for collecting fewer files. This should help a bit without resuming.

**Comment by fr4gment** [comment]
I appreciate the details and the improvements in 0.73 sound awesome and I think that will help similar problems and provide some good QoL improvements. My suggestion, though, was not necessarily an automated one but a manual one by a user. 

They can still reassess whether there is a better method, but doesn't cause a large upload to completely lose all of its progress if that is indeed the route they need to take.

So I think I'd want the best of both worlds :D. Still have the upload depend on the collection (allowing users to interrupt it if necessary) but also giving them a mechanism that if the upload failed, they are able to "restart" it and it would retain *most* of the progress it's already made.

**Comment by scudette** [comment]
I dont have a problem with a manual intervention mechanism for resuming uploads in some cases. It will have to be a user judgement call but will be useful.

We can already almost do that now:
1. Write an artifact with glob that emits a list of all the upload that it wants to upload first, these will hit the server first and should be recorded in one artifact source. For example KapeFiles.Target will usually have the `All File Metadata` source sent first anyway.
2. As uploads are sent the `Uploads` source is being filled up.
3. If the artifact fails, we can write a VQL query can just look at the files that the artifact wanted to upload and remove the files that the artifact managed to upload and reupload those ones.

We can do it in a more seamless way but essentially this can be done as a manual step - if we can re-upload a partial set we can save time. Maybe we can work on a POC like that for Kapefiles.Target and standardize the way to do it for different artifacts so it is more seamless.

**Comment by fr4gment** [comment]
I agree the Kapefiles.Target seems like a good use case to design towards. Just to emphasize, I'd want to extend it to the individual files themselves. 

As I mentioned, there are times where we need to pull large files off of host (in some cases many GBs). Having the upload maintain progress on such a file and allowing a user to tell velo to continue from its last "checkpoint" would be extremely useful. I've had experiences where I was uploading a large file that was needed (~40GB) and the upload would get to about 30 GB and then the client would crash or it would stall out, requiring me to start the upload again from 0. I was never actually able to be finished after several tries and quite a bit of waiting, which is precious time to have during an investigation.

Kapefiles.Target may be subjected to large files and a large amount of files, depending on what is being collected, so I do agree this artifact would provide good material to design for.

**Comment by scudette** [comment]
Implemented by #4364
